pkg/cli/redis.go

```diff
@@ -0,0 +1,167 @@
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/pkg/errors"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+)
+
+const (
+	// RedisUserFlag is the ENV var for the Redis username
+	RedisUserFlag string = "redis-user"
+	// RedisPasswordFlag is the ENV var for the Redis password
+	RedisPasswordFlag string = "redis-password"
+	// RedisHostFlag is the ENV var for the Redis hostname
+	RedisHostFlag string = "redis-host"
+	// RedisPortFlag is the ENV var for the Redis port
+	RedisPortFlag string = "redis-port"
+	// RedisDBNameFlag is the ENV var for the Redis database name, which
+	// is represented by a positive integer. Using multiple databases in
+	// the same Redis instance allows separating concerns.
+	RedisDBNameFlag string = "redis-db-name"
+	// RedisConnectTimeoutFlag specifies how long to wait to establish a
+	// connection to the Redis instance
+	RedisConnectTimeoutFlag string = "redis-connect-timeout-in-seconds"
+	// RedisEnabledFlag specifies whether or not we attempt to connect
+	// to Redis. For example, apps that use mTLS don't need Redis.
+	RedisEnabledFlag string = "redis-enabled"
+)
+
+// InitRedisFlags initializes RedisFlags command line flags
+func InitRedisFlags(flag *pflag.FlagSet) {
+	flag.String(RedisUserFlag, "", "Redis username")
+	flag.String(RedisPasswordFlag, "", "Redis password")
+	flag.String(RedisHostFlag, "localhost", "Redis hostname")
+	flag.Int(RedisPortFlag, 6379, "Redis port")
+	flag.Int(RedisDBNameFlag, 0, "Redis database")
+	flag.Int(RedisConnectTimeoutFlag, 2, "Redis connect timeout in seconds")
+	flag.Bool(RedisEnabledFlag, true, "Whether or not Redis is enabled")
+}
+
+// CheckRedis validates Redis command line flags
+func CheckRedis(v *viper.Viper) error {
+	enabled := v.GetBool(RedisEnabledFlag)
+	if !enabled {
+		return nil
+	}
+
+	host := v.GetString(RedisHostFlag)
+	fmt.Println("Redis host is:", host)
+	if err := ValidatePort(v, RedisPortFlag); err != nil {
+		return err
+	}
+
+	environment := v.GetString(EnvironmentFlag)
+	if environment == EnvironmentProd {
+		if err := ValidateRedisHost(v, RedisHostFlag); err != nil {
+			return err
+		}
+	}
+
+	if err := ValidateRedisConnectTimeout(v, RedisConnectTimeoutFlag); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// InitRedis initializes a Redis pool from command line flags.
+// v is the viper Configuration.
+// logger is the application logger.
+func InitRedis(v *viper.Viper, logger Logger) (*redis.Pool, error) {
+	enabled := v.GetBool(RedisEnabledFlag)
+	if !enabled {
+		return nil, nil
+	}
+
+	redisUser := v.GetString(RedisUserFlag)
+	redisPassword := v.GetString(RedisPasswordFlag)
+	redisHost := v.GetString(RedisHostFlag)
+	redisPort := strconv.Itoa(v.GetInt(RedisPortFlag))
+	redisDBName := v.GetString(RedisDBNameFlag)
+	redisConnectTimeout := v.GetInt(RedisConnectTimeoutFlag)
+	timeoutDuration := time.Duration(redisConnectTimeout) * time.Second
+
+	s := "redis://%s:%s@%s:%s/%s"
```

Comment:
It turns out that username is optional. When using the Auth Token mechanism which is employed by AWS Elasticache we are only using a password. That would modify the auth here to look like:

```suggestion
	s := "redis://%s@%s:%s/%s"
```

So I'd add in a check here and if the username, password, or both are missing then the uri template should be changed.

Also, can you modify the variable name to be `redisUriTemplate` instead of `s`?